components/hab/src/main.rs

```diff
@@ -901,9 +906,37 @@ fn sub_pkg_bulkupload(ui: &mut UI, m: &ArgMatches<'_>) -> Result<()> {
                       &artifact_path.display(),
                       key_path.display()))?;
     ui.status(Status::Found,
-              format!("{} artifacts for upload.", artifact_paths.len()))?;
-
-    for artifact_path in artifact_paths {
+              format!("{} artifact(s) for upload.", artifact_paths.len()))?;
+
+    if create_origins {
+        let mut origins: Vec<String> = Vec::new();
+        ui.status(Status::Discovering,
+                  "origin names from artifact metadata..".to_string())?;
+        for artifact_path in &artifact_paths {
+            let ident = PackageArchive::new(&artifact_path).ident()?;
+            origins.push(ident.origin);
+        }
+        origins.sort();
+        origins.dedup();
+        for origin in &origins {
+            ui.status(Status::Custom(Glyph::CheckMark, "".to_string()), origin)?;
+        }
+        ui.status(Status::Found,
+                  format!("{} origin(s) that may need to be created.", origins.len()))?;
```

Comment:
I think we should only give this warning if we find origins that are NOT present.  This seems like a generic warning being given all the time?

Reply:
Agreed that is less useful then what I had intended.
Because the Builder API [origins](https://github.com/habitat-sh/builder/blob/1626c3eec4389329df9044ae5a9075c0faba89e1/components/builder-api/doc/api.raml#L780) endpoint doesn't currently implement a `GET` or list functionality to enumerate the present origins, we just have to assume none of the origins exist and call `command::origin::create::start()` on each. 
Due to all of the above, I've just removed the message entirely in https://github.com/habitat-sh/habitat/pull/7133/commits/0c96916156ce8d0eeb4636319945c48c42d56bc6

Reply:
It seems that I am mistaken. The Builder API **does** implement a `GET` response on an origin name. 

**Note**: It is slightly more expensive to implement a `GET` followed by a `POST` as opposed to just blindly issuing a `POST` for the same end result. The existence check also adds some minor logic complexity as an additional cost.

The question then becomes: Is the additional existence check worthwhile to be able to provide the UX of only displaying the origins that need to be created?

Reply:
I think it might be worthwhile to do the extra work, and change the flow a little bit.

Consider the case of a sync after the initial bootstrap; all the origins will be in place (most likely). I think the user would like to be warned if a new origin was needed, rather than have to mentally go through and say 'ok, core is ok, chef is ok, etc'.

If I understand the current behavior, if create_origins is false, and and an origin does not exist, we will go through the upload process until we try to upload a package to a non-existent origin, where we will then error. Given that the user's attention might have shifted to something else while waiting for uploads, that can create unnecessary delay and friction. 

Maybe we should instead always check the origins, and list missing origins. If create_origins is true, we go on and offer to create them, and if it's false we fail early. 

The main downside is the walk through the artifacts path and the 'GET' from builder will impose some extra work. I have some ideas for optimizations if that turns out to be expensive.

Reply:
@markan @apriofrost @chefsalim with this [commit](https://github.com/habitat-sh/habitat/pull/7133/commits/9f726c1fe1d6a3a53fa47c00d9f2ef5340de6c0a) there is only one CLI argument related to origin creation ("--auto-create-origins"). That is in alignment with @apriofrost's UX [diagram](https://github.com/habitat-sh/habitat/pull/7133#pullrequestreview-310218276) above. I believe that changes some of your reasoning above.

It also ends up meaning that origins are always auto created, the only difference is whether or not we prompt the user for confirmation.

Keeping in mind that `ommand::origin::create::start()` on each origin issues a `POST` but handles any existing origins gracefully:

```
Ω Creating origin chef.
… Skipping creation of origin chef. Origin already exists!
```

Due to the fact that in the current implementation origins are _always_ created and the creation method is effectively idempotent, is it worth it to implement a check?

Reply:
I think I'll go ahead and implement an existence check for the sake of completeness and elimination of any assumptions 😃

Reply:
Thanks for the clarification of the behavior!

I still lean towards the existence check. I'm thinking about the sync case, where most of time the origins will exist. If we list those every time, even the ones that exist already, it will train the user to hit 'yes' (or set auto-create). If a new origin appears the chances are that reflex will have them say 'yes' because that's become the pattern. It's possible they will create an origin they didn't  intend to that way.

Reply:
I've implemented the client for the origin API check in https://github.com/habitat-sh/habitat/pull/7133/commits/bc818e341240f108e0210ce9c7c27bd80e522fed